src/Kernel/Process.class.st

```diff
@@ -145,23 +145,20 @@ Process >> calleeOf: aContext [
 ]
 
 { #category : #'changing suspended state' }
-Process >> complete: aContext [ 
+Process >> complete: aContext [
 	"Run self until aContext is popped or an unhandled error is raised.  Return self's new top context, unless an unhandled error was raised then return the signaler context (rather than open a debugger)."
-	
+
 	| ctxt pair error |
 	ctxt := suspendedContext.
-	suspendedContext := nil.  "disable this process while running its stack in active process below"
+	suspendedContext := nil.	"disable this process while running its stack in active process below"
 	pair := Processor activeProcess
-				evaluate: [ctxt runUntilErrorOrReturnFrom: aContext]
+				evaluate: [ ctxt runUntilErrorOrReturnFrom: aContext ]
 				onBehalfOf: self.
 	suspendedContext := pair first.
 	error := pair second.
-	error class = Halt ifTrue: [ 
-		aContext methodNode 
-			pragmaNamed: #debuggerCompleteToSender 
-			ifPresent: [ ^ self completeTo: aContext sender ] ].
-	error ifNotNil: [^ error signalerContext].
-	^ suspendedContext
+	error ifNil: [ ^ suspendedContext ].
+	(error handleProcessCompletionIfNeeded: self with: aContext) ifNotNil: [ :result | ^ result ].
```

Comment:
it's nice. But let's improve it to avoid nil result logic.
By default #handle method will return "self signallerContext" and Halt will do extra logic.

Reply:
And maybe method name can be simpler and follow sender name #complete:
```Smalltalk
error completeProcess:  self at: aContext
```